tool/planet/units.go

```diff
@@ -0,0 +1,116 @@
+package main
+
+import (
+	"context"
+
+	"github.com/gravitational/planet/lib/utils"
+
+	systemdDbus "github.com/coreos/go-systemd/dbus"
+	"github.com/gravitational/trace"
+	log "github.com/sirupsen/logrus"
+)
+
+func newUnitMonitor(units ...string) (*unitMonitor, error) {
+	conn, err := systemdDbus.NewSystemConnection()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &unitMonitor{
+		conn:  conn,
+		units: units,
+	}, nil
+}
+
+type unitMonitor struct {
+	conn  *systemdDbus.Conn
+	units []string
+}
+
+func (r *unitMonitor) close() {
+	r.conn.Close()
+}
+
+func (r *unitMonitor) start(ctx context.Context) error {
+	resultC := make(chan string, len(r.units))
+	// Number of successfully started start jobs
+	var numStarted int
+	for _, unit := range r.units {
+		logger := log.WithField("unit", unit)
+		logger.Info("Start service.")
+		if _, err := r.conn.StartUnit(unit, "replace", resultC); err != nil {
+			logger.WithError(err).Warn("Failed to start unit.")
+			continue
+		}
+		numStarted += 1
+	}
+	if numStarted == 0 {
+		return nil
+	}
+	r.waitForJobs(ctx, resultC, numStarted)
```

Comment:
Overall, I'm not sure I understand of waiting for the jobs resultC is? It only seems to block execution, so is the idea to not try and send too many commands to systemd is things are flapping?

Reply:
Yes, that's the idea. Usually the notifications are pretty snappy but who knows.